internal/tuf/v01/root.go

```diff
@@ -784,3 +786,108 @@ func (o *OtherRepository) GetInitialRootPrincipals() []tuf.Principal {
 	}
 	return initialRootPrincipals
 }
+
+// Hooks defines the schema that groups hooks by stage.
+type Hooks struct {
+	PreCommit map[string]Hook `json:"pre-commit"`
+	PrePush   map[string]Hook `json:"pre-push"`
+}
+
+// AddHook adds the specified hook to the metadata.
+func (r *RootMetadata) AddHook(stage, hookName, env string, hashes map[string]gitinterface.Hash, modules, principalIDs []string) error {
+	// TODO: Check if principal actually exists in TargetsMetadata?
+
+	newHook := Hook{
+		Name:         hookName,
+		PrincipalIDs: set.NewSetFromItems(principalIDs...),
+		Hashes:       hashes,
+		Environment:  env,
+		Modules:      modules,
+	}
+
+	switch stage {
+	case "pre-commit":
+		if r.Hooks.PreCommit == nil {
+			r.Hooks.PreCommit = make(map[string]Hook, 1)
+		}
+		r.Hooks.PreCommit[hookName] = newHook
+	case "pre-push":
+		if r.Hooks.PrePush == nil {
+			r.Hooks.PrePush = make(map[string]Hook, 1)
+		}
+		r.Hooks.PrePush[hookName] = newHook
+	default:
+		return tuf.ErrInvalidHookStage
+	}
+
+	return nil
+}
+
+// RemoveHook removes the hook specified by stage and hookName.
+func (r *RootMetadata) RemoveHook(stage, hookName string) error {
+	switch stage {
+	case "pre-commit":
+		delete(r.Hooks.PreCommit, hookName)
+	case "pre-push":
+		delete(r.Hooks.PrePush, hookName)
+	default:
+		return tuf.ErrInvalidHookStage
+	}
+
+	return nil
+}
+
+// GetHooks returns the hooks for the specified stage.
+func (r *RootMetadata) GetHooks(stage string) (map[string]tuf.Applet, error) {
+	var originMap map[string]Hook
+	switch stage {
+	case "pre-commit":
+		originMap = r.Hooks.PreCommit
+	case "pre-push":
+		originMap = r.Hooks.PrePush
+	default:
+		return nil, tuf.ErrInvalidHookStage
+	}
+
+	var appletMap = make(map[string]tuf.Applet)
+
+	for name, hook := range originMap {
+		appletMap[name] = &hook
+	}
+
+	return appletMap, nil
+}
+
+// Hook defines the schema for a hook.
+type Hook struct {
+	Name         string                       `json:"name"`
+	PrincipalIDs *set.Set[string]             `json:"principals"`
+	Hashes       map[string]gitinterface.Hash `json:"hashes"`
+	Environment  string                       `json:"environment"`
+	Modules      []string                     `json:"modules"`
+}
+
+// ID returns the identifier of the hook, its name.
+func (h *Hook) ID() string {
+	return h.Name
+}
+
+// GetPrincipalIDs returns the principals that must run this hook.
+func (h *Hook) GetPrincipalIDs() *set.Set[string] {
+	return h.PrincipalIDs
+}
+
+// GetHashes returns the hashes of the hook file.
```

Comment:
Is this to enable hash agility for the hook file? Having `gitinterface.Hash` as a target type might be blocking if so.

Reply:
I would probably make it just a string, to store hex encoded representations. We should look to upstream TUF implementations here.